### ISSUE TITLE
Insufficient shared memory

### ISSUE THREAD

**Issue by ShawnNew** [microsoft/pai#4478]
When I want to submit a job that use pytorch to train, there will be an issue claim about no sufficient shared memory. 

I cannot also add `shmMB` into job config yaml. `ValidationError: child "taskRoles" fails because [child "Task_role_1" fails because [child "resourcePerInstance" fails because ["shmMB" is not allowed]]]` will be reported by validation check.
 
I am using openpai v0.15.0, any suggest about how to fix it?

**Comment by abuccts** [comment]
what's your config file

https://github.com/microsoft/pai/blob/17794fbe6a75b4fbf496e1011c6ef6d8c005c9a6/docs/pai-job-protocol.yaml#L60-L61

**Comment by qingdo** [comment]
When I added to the YAML file and closed it, this item disappear (extraContainerOptions: 
   shmMB: 4096). My version is 0.14.0.

I think there is another user met this issue. 
https://github.com/Microsoft/pai/issues/2656#issuecomment-573555422


My scripts is. 
```
protocolVersion: 2
name: xxx
type: job
jobRetryCount: 0
prerequisites:
  - type: dockerimage
    uri: 'ufoym/deepo:pytorch-py36-cu90'
    name: docker_image_0
taskRoles:
  Task_role_1:
    instances: 1
    completion:
      minFailedInstances: 1
      minSucceededInstances: 1
    dockerImage: docker_image_0
    extraContainerOptions:
      shmMB: 4096
    resourcePerInstance:
      gpu: 8
      cpu: 32
      memoryMB: 65536
    commands:
      - echo 'done'
    taskRetryCount: 0
defaults:
  virtualCluster: default
```

**Comment by ShawnNew** [comment]
Hi @abuccts, 

My config file looks like this:
![image](https://user-images.githubusercontent.com/21559339/81140579-69262880-8f9c-11ea-9776-cb11f3a2082a.png)
I haven't change anything of this file.

**Comment by abuccts** [comment]
I mean, what's job config file. you said you submitted a pytorch job, what did you submit?

**Comment by ShawnNew** [comment]
The file looks like the following, it's just a simple command `df -h` to output shm.
```
protocolVersion: 2
name: 1588225242282_5df02a7e
type: job
jobRetryCount: 0
prerequisites:
  - type: dockerimage
    uri: 'dl-dev:latest'
    name: docker_image_0
taskRoles:
  Task_role_1:
    instances: 1
    completion:
      minFailedInstances: 1
      minSucceededInstances: 1
    dockerImage: docker_image_0
    resourcePerInstance:
      gpu: 1
      cpu: 4
      memoryMB: 8192
    commands:
      - '`#TEAMWISE_STORAGE_START`'
      - '`#Auto generated code, please do not modify`'
      - apt-get update
      - apt-get install --assume-yes nfs-common cifs-utils sshpass wget
      - umask 000
      - '`#TEAMWISE_STORAGE_END`'
      - cd /user/code/intention-prediction
      - df -h
    taskRetryCount: 0
defaults:
  virtualCluster: default
```

**Comment by ShawnNew** [comment]
I tried to add fileds `extraContainerOptions` and child field `shmMB`, the validation check is passed. But the added fields will not show up when I close and re-open the job config file.

**Comment by abuccts** [comment]
seems like a frontend issue, if you clicked "Save" button and it didn't work, there're two options:

* save the above config to a yaml file `job.yaml` and submit through rest api

    ```sh
    curl -H "Content-Type: application/x-www-form-urlencoded" \
         -X POST http://MASTER_IP/rest-server/api/v1/authn/basic/login \
         -d "username=YOUR_USERNAME" -d "password=YOUR_PASSWORD"

    curl -H "Content-Type: text/yaml" \
         -H "Authorization: Bearer ABOVE_TOKEN" \
         -X POST http://MASTER_IP/rest-server/api/v2/jobs \
         --data-binary @job.yaml
    ```

* or update to a newer version like v0.17

**Comment by ShawnNew** [comment]
Ok, I'd like to upgrade to v0.17 and will post here if other problems happen.

**Comment by ShawnNew** [comment]
Hi @abuccts,

I upgraded to v0.17 and can save `shmMB` filed to config file. However, in the webportal the job succeeded without any output. There is `Not Found` in stdout.

Another problem of v0.17 is I cannot add nfs use storage_plugin the same way as v0.14. Can you please show me the correct way?

**Comment by fanyangCS** [comment]
For YARN version, please stick to 0.14. For version higher than 0.14, we have not tested the YARN version. We will release a pure k8s version soon and will provide some document to show how to migrate from YARN version to k8s version.

**Comment by abuccts** [comment]
> I upgraded to v0.17 and can save shmMB field to config file. However, in the webportal the job succeeded without any output. There is Not Found in stdout.

there's [an issue](https://github.com/microsoft/pai/issues/4306#issuecomment-601555216) for multi-line commands, you could try to switch off storage plugin on webportal and use `&&` to concatenate commands, e.g. `cd /user/code/intention-prediction && df -h`

> Another problem of v0.17 is I cannot add nfs use storage_plugin the same way as v0.14

this may also related to the multi-line commands, as a workaround, you could try to switch off storage plugin and concatenate the auto-generated commands yourself

---

as mentioned above, we haven't maintained yarn version since v0.14 so there exist some regression issues, it's recommended to migrate to pure k8s version

**Comment by ShawnNew** [comment]
To conclude, I have two problems as follow:

* I tested to issue a job of just one line command: `df -h` and there is no stdout. Really don't know how to debug this :crying_cat_face: since the job is showed succeed in the webportal.

* For nfs storage, I am a little bit confused of how to use that. I previously succeed mount external nfs server to openpai by using storage_plugin in v0.14 but cannot do the same thing in v0.17. And there are different ways in the document.

**Comment by ShawnNew** [comment]
Does anyone knows how to enable `stdout` in v0.17 just like v0.14?

**Comment by Binyang2014** [comment]
Hi @ShawnNew, v0.17 is an internal release. This internal release is pre-release for pure k8s version. For yarn version, we don't fully tested and unstable.

If you don't so urgent, we recommend you wait for a couple of weeks. The 1.0 version will be ready soon. This is a pure k8s version, and yarn version is deprecated.

For yarn version, we recommend stick to v0.14. Yarn version in PAI which above v0.14 is not stable, and not fully tested. We are not sure if you will encounter other bugs if you still use yarn version.

If you really want to enable `stdout` in v0.17 for yarn version.
Please provide the log from ![image](https://user-images.githubusercontent.com/9415966/77139022-49e84000-6aaf-11ea-815b-ea011cb68de7.png)
It should be `Go to yarn tracking page` for yarn version.

**Comment by ShawnNew** [comment]
Hi @Binyang2014,

Thanks for the reply, I will save std stream to local file and check the process.

Looking forward to 1.0 version.

**Comment by Binyang2014** [comment]
@ShawnNew 1.0 is released, Please try 1.0 version. Will close this issue
